modules/3d/src/ptcloud/sampling.cpp

```diff
@@ -278,30 +245,88 @@ int farthestPointSampling(OutputArray sampled_point_flags, InputArray input_pts,
     }
 
     // Randomly take a seed point from C and put it into S
-    int seed = (int)((rng? rng->next(): theRNG().next()) % ori_pts_size);
+    int seed = (int) ((rng ? rng->next() : theRNG().next()) % ori_pts_size);
     idxs[0] = seed;
     idxs[seed] = 0;
 
-    std::vector<char> _sampled_point_flags(ori_pts_size, 0);
-    _sampled_point_flags[seed] = 1;
+    // Pointer (base address) of access point data x,y,z
+    float *const ori_pts_ptr_x = (float *) ori_pts.data;
+    float *const ori_pts_ptr_y = ori_pts_ptr_x + ori_pts_size;
+    float *const ori_pts_ptr_z = ori_pts_ptr_y + ori_pts_size;
+
+#ifdef CV_SIMD
+    // Ensure that the point(in C) data x,y,z is continuous in the memory respectively
+    _swap(ori_pts_ptr_x[seed], ori_pts_ptr_x[0]);
+    _swap(ori_pts_ptr_y[seed], ori_pts_ptr_y[0]);
+    _swap(ori_pts_ptr_z[seed], ori_pts_ptr_z[0]);
+#endif
 
-    float *const ori_pts_ptr = (float *) ori_pts.data;
     int sampled_cnt = 1;
     const float dist_lower_limit_square = dist_lower_limit * dist_lower_limit;
     while (sampled_cnt < sampled_pts_size)
     {
+        // The position of the point that was last added to S
+#ifdef CV_SIMD
         int last_pt = sampled_cnt - 1;
-        float *const last_pt_ptr_base = ori_pts_ptr + 3 * idxs[last_pt];
-        float last_pt_x = last_pt_ptr_base[0], last_pt_y = last_pt_ptr_base[1], last_pt_z = last_pt_ptr_base[2];
+#elif
+        int last_pt = idxs[sampled_cnt - 1];
+#endif
```

Comment:
Usually SIMD optimization should not change meaning of used arrays / variables.
Otherwise similar changes can be applied to generic C++ code.

SIMD code should mirror behavior of C++ code to prove their equivalence.

---

If you need some extra requirements for SIMD code, like alignment or contiguous packing then do NOT use the same variable names.

Reply:
Alright, got it. The new code will maintain a uniform computing style.